velox/exec/prefixsort/tests/PrefixEncoderTest.cpp

```diff
@@ -0,0 +1,214 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <glog/logging.h>
+#include <gtest/gtest.h>
+
+#include "velox/exec/prefixsort/PrefixSortEncoder.h"
+
+namespace facebook::velox::exec::prefixsort::test {
+
+class PrefixEncoderTest : public testing::Test {
+ public:
+  template <typename T>
+  void testEncodeNoNull(T value, char* expectedAsc, char* expectedDesc) {
+    char encoded[sizeof(T)];
+    ascNullsFirstEncoder_.encodeNoNulls(value, (char*)encoded);
+    ASSERT_EQ(std::memcmp(encoded, expectedAsc, sizeof(T)), 0);
+    descNullsFirstEncoder_.encodeNoNulls(value, (char*)encoded);
+    ASSERT_EQ(std::memcmp(encoded, expectedDesc, sizeof(T)), 0);
+  }
+
+  template <typename T>
+  void testEncodeWithNull(T testValue, char* expectedAsc, char* expectedDesc) {
+    std::optional<T> nullValue = std::nullopt;
+    std::optional<T> value = testValue;
+    char encoded[sizeof(T) + 1];
+
+    ascNullsFirstEncoder_.encode(nullValue, encoded);
+    ASSERT_EQ(encoded[0], 0);
+    ascNullsLstEncoder_.encode(nullValue, encoded);
+    ASSERT_EQ(encoded[0], 1);
+    ascNullsFirstEncoder_.encode(value, encoded);
+    ASSERT_EQ(encoded[0], 1);
+    ASSERT_EQ(std::memcmp(encoded + 1, expectedAsc, sizeof(T)), 0);
+    ascNullsLstEncoder_.encode(value, encoded);
+    ASSERT_EQ(encoded[0], 0);
+    ASSERT_EQ(std::memcmp(encoded + 1, expectedAsc, sizeof(T)), 0);
+    descNullsFirstEncoder_.encode(value, encoded);
+    ASSERT_EQ(encoded[0], 1);
+    ASSERT_EQ(std::memcmp(encoded + 1, expectedDesc, sizeof(T)), 0);
+    descNullsLstEncoder_.encode(value, encoded);
+    ASSERT_EQ(encoded[0], 0);
+    ASSERT_EQ(std::memcmp(encoded + 1, expectedDesc, sizeof(T)), 0);
+  }
+
+  template <typename T>
+  void testEncode(T value, char* expectedAsc, char* expectedDesc) {
+    testEncodeNoNull<T>(value, expectedAsc, expectedDesc);
+    testEncodeWithNull<T>(value, expectedAsc, expectedDesc);
+  }
+
+  template <typename T>
+  void testNullCompare() {
+    std::optional<T> nullValue = std::nullopt;
+    std::optional<T> max = std::numeric_limits<T>::max();
+    std::optional<T> min = std::numeric_limits<T>::min();
+    char encodedNull[sizeof(T) + 1];
+    char encodedMax[sizeof(T) + 1];
+    char encodedMin[sizeof(T) + 1];
+
+    ascNullsFirstEncoder_.encode(nullValue, encodedNull);
+    ascNullsFirstEncoder_.encode(min, encodedMin);
+    // Null is less than min when asc && nullsFirst
+    ASSERT_EQ(std::memcmp(encodedNull, encodedMin, sizeof(T) + 1), -1);
+
+    descNullsFirstEncoder_.encode(nullValue, encodedNull);
```

Comment:
There is a lot of repetition in this code which makes it a bit hard to read. Consider using some lambdas:

```
    auto encode = [&] (auto& encoder) {
      encoder.encode(nullValue, encodedNull);
      encoder.encode(min, encodedMin);
      encoder.encode(max, encodedMax);
    };

    auto compare = [&] (const auto& left, const auto& right) {
      return std::memcmp(left, right, sizeof(T) + 1);
    };

    encode(ascNullsFirstEncoder_);

    // Nulls first: NULL < min < max.
    ASSERT_LT(compare(encodedNull, encodedMin), 0);
    ASSERT_LT(compare(encodedNull, encodedMax), 0);
    ASSERT_LT(compare(encodedMin, encodedMax), 0);

    // x == x.
    ASSERT_EQ(compare(encodedNull, encodedNull), 0);
    ASSERT_EQ(compare(encodedMin, encodedMin), 0);
    ASSERT_EQ(compare(encodedMax, encodedMax), 0);

    encode(descNullsFirstEncoder_);

    ASSERT_LT(compare(encodedNull, encodedMin), 0);
    ASSERT_LT(compare(encodedNull, encodedMax), 0);
    ASSERT_GT(compare(encodedMin, encodedMax), 0);

    // Nulls last: min < max < NULL.
    encode(ascNullsLstEncoder_);

    ASSERT_GT(compare(encodedNull, encodedMin), 0);
    ASSERT_GT(compare(encodedNull, encodedMax), 0);
    ASSERT_LT(compare(encodedMin, encodedMax), 0);

    encode(descNullsLastEncoder_);

    ASSERT_GT(compare(encodedNull, encodedMin), 0);
    ASSERT_GT(compare(encodedNull, encodedMax), 0);
    ASSERT_GT(compare(encodedMin, encodedMax), 0);
```